### DISCUSSION TITLE
async onTransition

### DISCUSSION THREAD

**Discussion by yarinvak** [statelyai/xstate#2444] [Q&A]
Hey all
I've got a question regarding async `onTransition`.

Let's say I have a state machine where some of the states are invoked in an async manner, and I interpret the machine using an async `onTransition` listener:
```
    interpretedMachine = interpret(machine).onTransition(async (state) => {
                ...
```

Inside the async lambda I'm persisting the state's context in DB. Now, I've noticed that the state machine transitions does not wait for the `onTransition` to resolve, meaning that the state machine can keep transitioning and invoking before the state is persisted in DB, which may cause some problem in my case.

Do you have any suggestion how to delay the state machine from being transitioned until the `onTransition` is resolved?

**Comment by davidkpiano** [comment]
Unfortunately for your use-case, that's not how state machines work. State machines are deterministic and transitions are conceptually "zero-time", so when an event happens, it immediately transitions.

It's sort of like asking for the weather to not rain yet until you get your umbrella. Nature doesn't work like that; it's going to rain on its own schedule no matter what. Same with state machines; it's going to transition when it receives an event, and it can't "wait" for some external event. `.onTransition()` is meant to be a _reaction_ to a state transition that _already occurred_.

-----

With your use-case though, there are other patterns you can use to accomplish the same behavior. For example, you can push state to a FIFO queue:

```js
const states = [];

interpret(machine).onTransition(state => states.push(state));

await new Promise(res => {
  while (states.length) {
    await persistToDB(states.pop());
  }
  res();
});
```

If you have a more full code example, I can provide other suggestions.
